test/Magnet_test.cpp

```diff
@@ -1,10 +1,195 @@
 #include "catch.hpp"
 #include "Magnet.h"
 
-TEST_CASE("Magnets initialise to the correct size")
+TEST_CASE("Magnets initialise to the correct size when done randomly")
 {
     Magnet magnet(10,20);
-    
     REQUIRE(magnet.getHeight() == 10);
     REQUIRE(magnet.getWidth() == 20);
 }
+
+TEST_CASE("Magnet constructor taking spins works")
+{
+    Magnet magnet({{1,-1,1},{-1,1,-1}});
+    std::vector<std::vector<int>> expected = {{1,-1,1},{-1,1,-1}};
+    REQUIRE(magnet.getSpins() == expected);
+}
+
+TEST_CASE("Magnet's update function leaves spins alone when necessary")
+{
+    std::vector<std::vector<int>> initialSpins{
+            { 1,-1, 1, 1},
+            { 1, 1, 1,-1},
+            {-1, 1,-1,-1},
+            {-1, 1, 1, 1}
+    };
+    Magnet magnet(initialSpins);
+
+    SECTION("Top-left corner")
+    {
+        REQUIRE(!magnet.update(0,0));
+        REQUIRE(magnet.getSpins() == initialSpins);
+    }
+    SECTION("Top-right corner")
+    {
+        REQUIRE(!magnet.update(0,magnet.getWidth()-1));
+        REQUIRE(magnet.getSpins() == initialSpins);
+    }
+    SECTION("Bottom-left corner")
+    {
+        REQUIRE(!magnet.update(magnet.getHeight()-1,0));
+        REQUIRE(magnet.getSpins() == initialSpins);
+    }
+    SECTION("Bottom-right corner")
+    {
+        REQUIRE(!magnet.update(magnet.getHeight()-1,magnet.getWidth()-1));
+        REQUIRE(magnet.getSpins() == initialSpins);
+    }
+    SECTION("Top edge")
+    {
+        REQUIRE(!magnet.update(0,2));
+        REQUIRE(magnet.getSpins() == initialSpins);
+    }
+    SECTION("Right edge")
+    {
+        REQUIRE(!magnet.update(2,3));
+        REQUIRE(magnet.getSpins() == initialSpins);
+    }
+    SECTION("Bottom edge")
+    {
+        REQUIRE(!magnet.update(3,2));
+        REQUIRE(magnet.getSpins() == initialSpins);
+    }
+    SECTION("Left edge")
+    {
+        REQUIRE(!magnet.update(1,0));
+        REQUIRE(magnet.getSpins() == initialSpins);
+    }
+    SECTION("Interior element")
+    {
+        REQUIRE(!magnet.update(1,2));
+        REQUIRE(magnet.getSpins() == initialSpins);
+    }
+}
+
+TEST_CASE("Magnet's update function updates when necessary")
+{
+    std::vector<std::vector<int>> initialSpins{
+            { 1,-1,-1, 1},
+            {-1, 1, 1,-1},
+            { 1,-1,-1, 1},
+            {-1, 1, 1,-1}
+    };
+    Magnet magnet(initialSpins);
+
+    SECTION("Top-left corner")
+    {
+        REQUIRE(magnet.update(0,0));
+        std::vector<std::vector<int>> expected{
+            {-1,-1,-1, 1},
+            {-1, 1, 1,-1},
+            { 1,-1,-1, 1},
+            {-1, 1, 1,-1}
+        };
```

Comment:
It could be more clear in the following test cases to just have a single deep-copy of the main magnet outside of the SECTIONs and manually change the entry we're interested in for this copy and require this to be equal to the updated entry. It might save some space and make things clearer, but I'm not sure about that, nor that it is helpful, maybe it is even unhelpful.

Reply:
I very much considered doing just what you suggested, however I settled for explicitly naming the result since I thought that would be slightly clearer here.  It's definitely at the cost of concise tests though - they now take up loads of room.  I don't have a hugely strong opinion either way though, and it sounds like you don't either, so I'm leaning on the side of leaving it as-is.

Reply:
I agree with that.